File: Tests/ClaudeUsageWidgetCoreTests/UsageStoreTests.swift
import Foundation
import Testing
@testable import ClaudeUsageWidgetCore

@MainActor
@Suite("UsageStore")
struct UsageStoreTests {
    static let now = Date(timeIntervalSince1970: 1_785_348_000)

    private static func snapshot(_ utilization: Double) -> UsageSnapshot {
        UsageSnapshot(buckets: ["five_hour": UsageBucket(utilization: utilization, resetsAt: nil)])
    }

    private func store(
        token: String? = "token",
        fetch: @escaping @Sendable (String) async throws -> UsageSnapshot
    ) -> UsageStore {
        let fixedNow = Self.now
        return UsageStore(fetch: fetch, tokenProvider: { token }, now: { fixedNow })
    }

    @Test("starts out loading")
    func startsLoading() {
        let store = store { _ in await Self.snapshot(1) }
        #expect(store.state == .loading)
    }

    @Test("a successful load publishes the snapshot and the fetch time")
    func publishesSnapshot() async {
        let store = store { _ in await Self.snapshot(42) }
        await store.load()
        #expect(store.state == .ok(Self.snapshot(42), fetchedAt: Self.now))
        #expect(store.lastSnapshot == Self.snapshot(42))
    }

    @Test("a missing token fails without calling the endpoint")
    func failsWithoutToken() async {
        let store = store(token: nil) { _ in
            Issue.record("fetch must not run without a token")
            return await Self.snapshot(1)
        }
        await store.load()
        #expect(store.state == .failed(.noCredentials))
    }

    @Test("a rejected token surfaces as unauthorized")
    func surfacesUnauthorized() async {
        let store = store { _ in throw UsageError.unauthorized }
        await store.load()
        #expect(store.state == .failed(.unauthorized))
    }

    @Test("an unexpected error surfaces as a network error")
    func wrapsUnknownErrors() async {
        struct Boom: Error {}
        let store = store { _ in throw Boom() }
        await store.load()
        if case .failed(.network) = store.state {} else {
            Issue.record("expected a network failure, got \(store.state)")
        }
    }

    @Test("a failure after a success keeps the last snapshot on screen")
    func keepsLastSnapshotOnFailure() async {
        final class Box: @unchecked Sendable { var shouldFail = false }
        let box = Box()
        let store = store { _ in
            if box.shouldFail { throw UsageError.unauthorized }
            return await Self.snapshot(42)
        }

        await store.load()
        box.shouldFail = true
        await store.load()

        #expect(store.state == .failed(.unauthorized))
        #expect(store.lastSnapshot == Self.snapshot(42))
    }

    @Test("the token is passed through to the fetch")
    func passesToken() async {
        final class Box: @unchecked Sendable { var seen: String? }
        let box = Box()
        let store = store(token: "sk-ant-oat01-example") { token in
            box.seen = token
            return await Self.snapshot(1)
        }
        await store.load()
        #expect(box.seen == "sk-ant-oat01-example")
    }

    @Test("a rate limit with a Retry-After pauses polling until the deadline plus a safety margin")
    func pausesOnRateLimit() async {
        final class Box: @unchecked Sendable { var calls = 0 }
        let box = Box()
        let store = store { _ in
            box.calls += 1
            throw UsageError.rateLimited(retryAfterSeconds: 600)
        }

        await store.load()
        #expect(store.state == .failed(.rateLimited(retryAfterSeconds: 600)))
        #expect(store.retryPausedUntil == Self.now.addingTimeInterval(600 + UsageStore.retryMargin))

        await store.load()
        #expect(box.calls == 1)
    }

    @Test("polling resumes once the rate-limit deadline has passed")
    func resumesAfterRateLimitDeadline() async {
        final class Clock: @unchecked Sendable { var now = Date.distantPast }
        final class Box: @unchecked Sendable { var calls = 0 }
        let clock = Clock()
        clock.now = Self.now
        let box = Box()
        let store = UsageStore(
            fetch: { _ in
                box.calls += 1
                if box.calls == 1 { throw UsageError.rateLimited(retryAfterSeconds: 600) }
                return await Self.snapshot(7)
            },
            tokenProvider: { "token" },
            now: { clock.now }
        )

        await store.load()
        clock.now = Self.now.addingTimeInterval(600 + UsageStore.retryMargin + 1)
        await store.load()

        #expect(box.calls == 2)
        #expect(store.state == .ok(Self.snapshot(7), fetchedAt: clock.now))
        #expect(store.retryPausedUntil == nil)
    }

    @Test("a rate limit without a Retry-After still backs off")
    func rateLimitWithoutRetryAfterBacksOff() async {
        let store = store { _ in throw UsageError.rateLimited(retryAfterSeconds: nil) }
        await store.load()
        #expect(store.retryPausedUntil == Self.now.addingTimeInterval(300 + UsageStore.retryMargin))
    }

    @Test("consecutive rate limits escalate the pause, so a hair-trigger ban is not re-tripped forever")
    func consecutiveRateLimitsEscalate() async {
        final class Clock: @unchecked Sendable { var now = Date.distantPast }
        let clock = Clock()
        clock.now = Self.now
        let store = UsageStore(
            fetch: { _ in throw UsageError.rateLimited(retryAfterSeconds: 60) },
            tokenProvider: { "token" },
            now: { clock.now }
        )

        await store.load()
        // First 429: the ladder's floor (300, one timer tick) beats the
        // server's 60 — a shorter pause could not be acted on anyway.
        #expect(store.retryPausedUntil == clock.now.addingTimeInterval(300 + UsageStore.retryMargin))

        clock.now = clock.now.addingTimeInterval(1000)
        await store.load()
        // Second in a row: the ladder doubles past the server's 60.
        #expect(store.retryPausedUntil == clock.now.addingTimeInterval(600 + UsageStore.retryMargin))

        clock.now = clock.now.addingTimeInterval(1000)
        await store.load()
        #expect(store.retryPausedUntil == clock.now.addingTimeInterval(1200 + UsageStore.retryMargin))
    }

    @Test("a success resets the rate-limit escalation")
    func successResetsEscalation() async {
        final class Clock: @unchecked Sendable { var now = Date.distantPast }
        final class Box: @unchecked Sendable { var fail = true }
        let clock = Clock()
        clock.now = Self.now
        let box = Box()
        let store = UsageStore(
            fetch: { _ in
                if box.fail { throw UsageError.rateLimited(retryAfterSeconds: nil) }
                return await Self.snapshot(1)
            },
            tokenProvider: { "token" },
            now: { clock.now }
        )

        await store.load()                                   // 429 #1
        clock.now = clock.now.addingTimeInterval(1000)
        box.fail = false
        await store.load()                                   // success
        box.fail = true
        clock.now = clock.now.addingTimeInterval(1000)
        await store.load()                                   // 429 again
        // Back to the first step of the ladder, not the third.
        #expect(store.retryPausedUntil == clock.now.addingTimeInterval(300 + UsageStore.retryMargin))
    }

    @Test("a refresh while one is already in flight does not start a second fetch")
    func coalescesOverlappingRefreshes() async {
        final class Box: @unchecked Sendable { var calls = 0 }
        let box = Box()
        let store = store { _ in
            box.calls += 1
            try? await Task.sleep(for: .milliseconds(50))
            return await Self.snapshot(1)
        }

        async let first: Void = store.load()
        async let second: Void = store.load()
        _ = await (first, second)

        #expect(box.calls == 1)
    }
}

File: Tests/ClaudeUsageWidget.Core.Tests/UsageStoreTests.cs
namespace ClaudeUsageWidget.Core.Tests;

public class UsageStoreTests
{
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_785_348_000);

    /// Mutable clock for tests that need `now` to move between calls.
    private sealed class Clock
    {
        public DateTimeOffset Now;
    }

    private static UsageSnapshot Snapshot(double utilization) =>
        new(new Dictionary<string, UsageBucket> { ["five_hour"] = new(utilization, null) });

    private static UsageStore MakeStore(
        bool hasCredentials = true,
        Func<CancellationToken, Task<UsageSnapshot>>? fetch = null) =>
        new(
            fetch ?? (_ => Task.FromResult(Snapshot(1))),
            () => hasCredentials,
            () => Now,
            () => null,
            _ => { });

    [Fact]
    public void StartsLoading()
    {
        var store = MakeStore();
        Assert.Equal(new UsageState.Loading(), store.CurrentState);
    }

    [Fact]
    public async Task PublishesSnapshot()
    {
        var store = MakeStore(fetch: _ => Task.FromResult(Snapshot(42)));
        await store.LoadAsync();
        Assert.Equal(new UsageState.Ok(Snapshot(42), Now), store.CurrentState);
        Assert.Equal(Snapshot(42), store.LastSnapshot);
    }

    // Replaces Swift's failsWithoutToken/passesToken: there is no token to pass
    // through any more (the transport is always the web session), only a
    // boolean gate on whether a session exists at all.
    [Fact]
    public async Task NoCredentialsFailsWithoutFetch()
    {
        var called = false;
        var store = MakeStore(hasCredentials: false, fetch: _ =>
        {
            called = true;
            return Task.FromResult(Snapshot(1));
        });

        await store.LoadAsync();

        Assert.Equal(new UsageState.Failed(UsageError.NoCredentials), store.CurrentState);
        Assert.False(called);
    }

    [Fact]
    public async Task SurfacesUnauthorized()
    {
        var store = MakeStore(fetch: _ => throw new UsageException(UsageError.Unauthorized));
        await store.LoadAsync();
        Assert.Equal(new UsageState.Failed(UsageError.Unauthorized), store.CurrentState);
    }

    [Fact]
    public async Task WrapsUnknownErrors()
    {
        var store = MakeStore(fetch: _ => throw new InvalidOperationException("boom"));
        await store.LoadAsync();
        var failed = Assert.IsType<UsageState.Failed>(store.CurrentState);
        Assert.Equal(UsageErrorKind.Network, failed.Error.Kind);
    }

    [Fact]
    public async Task KeepsLastSnapshotOnFailure()
    {
        var shouldFail = false;
        var store = MakeStore(fetch: _ => shouldFail
            ? throw new UsageException(UsageError.Unauthorized)
            : Task.FromResult(Snapshot(42)));

        await store.LoadAsync();
        shouldFail = true;
        await store.LoadAsync();

        Assert.Equal(new UsageState.Failed(UsageError.Unauthorized), store.CurrentState);
        Assert.Equal(Snapshot(42), store.LastSnapshot);
    }

    [Fact]
    public async Task PausesOnRateLimit()
    {
        var calls = 0;
        var store = MakeStore(fetch: _ =>
        {
            calls++;
            throw new UsageException(UsageError.RateLimited(600));
        });

        await store.LoadAsync();
        Assert.Equal(new UsageState.Failed(UsageError.RateLimited(600)), store.CurrentState);
        Assert.Equal(Now + TimeSpan.FromSeconds(3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);

        await store.LoadAsync();
        Assert.Equal(1, calls);
    }

    [Fact]
    public async Task ResumesAfterRateLimitDeadline()
    {
        var clock = new Clock { Now = Now };
        var calls = 0;
        var store = new UsageStore(
            fetch: _ =>
            {
                calls++;
                if (calls == 1) throw new UsageException(UsageError.RateLimited(600));
                return Task.FromResult(Snapshot(7));
            },
            hasCredentials: () => true,
            now: () => clock.Now,
            loadRetryState: () => null,
            saveRetryState: _ => { });

        await store.LoadAsync();
        clock.Now = Now + TimeSpan.FromSeconds(3600 + UsageStore.RetryMarginSeconds + 1);
        await store.LoadAsync();

        Assert.Equal(2, calls);
        Assert.Equal(new UsageState.Ok(Snapshot(7), clock.Now), store.CurrentState);
        Assert.Null(store.RetryPausedUntil);
    }

    [Fact]
    public async Task RateLimitWithoutRetryAfterBacksOff()
    {
        var store = MakeStore(fetch: _ => throw new UsageException(UsageError.RateLimited(null)));
        await store.LoadAsync();
        Assert.Equal(Now + TimeSpan.FromSeconds(3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);
    }

    [Fact]
    public async Task ConsecutiveRateLimitsEscalate()
    {
        var clock = new Clock { Now = Now };
        var store = new UsageStore(
            fetch: _ => throw new UsageException(UsageError.RateLimited(60)),
            hasCredentials: () => true,
            now: () => clock.Now,
            loadRetryState: () => null,
            saveRetryState: _ => { });

        await store.LoadAsync();
        Assert.Equal(clock.Now + TimeSpan.FromSeconds(3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);

        clock.Now += TimeSpan.FromSeconds(4000);
        await store.LoadAsync();
        Assert.Equal(clock.Now + TimeSpan.FromSeconds(6 * 3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);

        clock.Now += TimeSpan.FromSeconds(22_000);
        await store.LoadAsync();
        Assert.Equal(clock.Now + TimeSpan.FromSeconds(24 * 3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);
    }

    [Fact]
    public async Task SuccessResetsEscalation()
    {
        var clock = new Clock { Now = Now };
        var fail = true;
        var store = new UsageStore(
            fetch: _ => fail
                ? throw new UsageException(UsageError.RateLimited(null))
                : Task.FromResult(Snapshot(1)),
            hasCredentials: () => true,
            now: () => clock.Now,
            loadRetryState: () => null,
            saveRetryState: _ => { });

        await store.LoadAsync();                           // 429 #1
        clock.Now += TimeSpan.FromSeconds(4000);
        fail = false;
        await store.LoadAsync();                            // success
        fail = true;
        clock.Now += TimeSpan.FromSeconds(1000);
        await store.LoadAsync();                            // 429 again

        // Back to the first step of the ladder, not the third.
        Assert.Equal(clock.Now + TimeSpan.FromSeconds(3600 + UsageStore.RetryMarginSeconds), store.RetryPausedUntil);
    }

    [Fact]
    public async Task PersistsRateLimitPause()
    {
        UsageRetryState? saved = null;
        var calls = 0;

        var first = new UsageStore(
            fetch: _ => throw new UsageException(UsageError.RateLimited(60)),
            hasCredentials: () => true,
            now: () => Now,
            loadRetryState: () => saved,
            saveRetryState: state => saved = state);

        await first.LoadAsync();
        Assert.NotNull(saved);

        var relaunched = new UsageStore(
            fetch: _ =>
            {
                calls++;
                return Task.FromResult(Snapshot(1));
            },
            hasCredentials: () => true,
            now: () => Now + TimeSpan.FromSeconds(30),
            loadRetryState: () => saved,
            saveRetryState: state => saved = state);

        await relaunched.LoadAsync();

        Assert.Equal(saved!.Until, relaunched.RetryPausedUntil);
        Assert.Equal(new UsageState.Failed(UsageError.RateLimited(null)), relaunched.CurrentState);
        Assert.Equal(0, calls);
    }

    [Fact]
    public async Task CoalescesOverlappingRefreshes()
    {
        var calls = 0;
        var store = MakeStore(fetch: async _ =>
        {
            calls++;
            await Task.Delay(50);
            return Snapshot(1);
        });

        var first = store.LoadAsync();
        var second = store.LoadAsync();
        await Task.WhenAll(first, second);

        Assert.Equal(1, calls);
    }

    // Changed has no Swift equivalent (Swift observation is implicit via
    // @Observable); these pin down the C#-specific event contract.
    [Fact]
    public async Task ChangedFiresOnSuccessfulLoad()
    {
        var store = MakeStore(fetch: _ => Task.FromResult(Snapshot(1)));
        var fireCount = 0;
        store.Changed += () => fireCount++;

        await store.LoadAsync();

        Assert.Equal(1, fireCount);
    }

    [Fact]
    public async Task ChangedDoesNotFireWhilePaused()
    {
        var store = MakeStore(fetch: _ => throw new UsageException(UsageError.RateLimited(600)));

        await store.LoadAsync(); // first failure: pauses and fires once
        var fireCount = 0;
        store.Changed += () => fireCount++;
        await store.LoadAsync(); // paused: returns without a state change

        Assert.Equal(0, fireCount);
    }
}
